pkg/connection/store/vault/client/client.go

```diff
@@ -0,0 +1,238 @@
+/*
+ Copyright 2022 The Crossplane Authors.
+
+ Licensed under the Apache License, Version 2.0 (the "License");
+ you may not use this file except in compliance with the License.
+ You may obtain a copy of the License at
+
+     http://www.apache.org/licenses/LICENSE-2.0
+
+ Unless required by applicable law or agreed to in writing, software
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+*/
+
+package client
+
+import (
+	"encoding/json"
+	"path/filepath"
+
+	"github.com/hashicorp/vault/api"
+
+	v1 "github.com/crossplane/crossplane-runtime/apis/common/v1"
+	"github.com/crossplane/crossplane-runtime/pkg/errors"
+	"github.com/crossplane/crossplane-runtime/pkg/resource"
+)
+
+const (
+	errGet           = "cannot get secret"
+	errDelete        = "cannot delete secret"
+	errRead          = "cannot read secret"
+	errWriteData     = "cannot write secret Data"
+	errWriteMetadata = "cannot write secret metadata Data"
+
+	// ErrNotFound is the error returned when secret does not exist.
+	ErrNotFound = "secret not found"
+)
+
+// KVSecret is a KVAdditiveClient Engine secret
+type KVSecret struct {
+	CustomMeta map[string]interface{}
+	Data       map[string]interface{}
+	version    json.Number
+}
+
+// LogicalClient is a client to perform logical backend operations on Vault.
+type LogicalClient interface {
+	Read(path string) (*api.Secret, error)
+	Write(path string, data map[string]interface{}) (*api.Secret, error)
+	Delete(path string) (*api.Secret, error)
+}
+
+// KVAdditiveOption configures a KVAdditiveClient.
+type KVAdditiveOption func(*KVAdditiveClient)
+
+// WithVersion specifies which version of KVAdditiveClient Secrets engine to be used.
+func WithVersion(v *v1.VaultKVVersion) KVAdditiveOption {
+	return func(kv *KVAdditiveClient) {
+		if v != nil {
+			kv.version = *v
+		}
+	}
+}
+
+// KVAdditiveClient is a Vault KV Secrets Engine client that adds new data
+// to existing ones while applying secrets.
+type KVAdditiveClient struct {
+	client    LogicalClient
+	mountPath string
+	version   v1.VaultKVVersion
+}
+
+// NewAdditiveClient returns a KVAdditiveClient.
+func NewAdditiveClient(logical LogicalClient, mountPath string, opts ...KVAdditiveOption) *KVAdditiveClient {
+	kv := &KVAdditiveClient{
+		client:    logical,
+		mountPath: mountPath,
+		version:   v1.VaultKVVersionV2,
+	}
+
+	for _, o := range opts {
+		o(kv)
+	}
+
+	return kv
+}
+
+// Get returns a KVSecret at a given path.
+func (k *KVAdditiveClient) Get(path string, secret *KVSecret) error {
+	dataPath := filepath.Join(k.mountPath, path)
+	if k.version == v1.VaultKVVersionV2 {
+		dataPath = filepath.Join(k.mountPath, "data", path)
+	}
```

Comment:
Would it make more sense to have separate clients for v1 and v2? It seems like most client functions have distinct branches to handle v1 vs v2 logic. Personally I could go either way.

Reply:
Yeah, this was also something I considered but ended up not doing as branching is still manageable and code ended up readable enough with early returns. I don't this client to grow too much, but we may consider separating later otherwise.

Reply:
Got linter warnings for gocyclo and ended up separating :)